core/metrics/src/main/scala-2/org/typelevel/otel4s/metrics/CounterMacro.scala

```diff
@@ -47,22 +68,37 @@ object CounterMacro {
   def add[A](c: blackbox.Context)(
       value: c.Expr[A],
       attributes: c.Expr[Attribute[_]]*
+  ): c.universe.Tree = {
+    import c.universe._
+    addColl(c)(value, c.Expr(q"_root_.scala.Seq(..$attributes)"))
```

Comment:
I think we can use `_root_.org.typelevel.otel4s.Attributes(..$attributes)` here. That way, we will have fewer conversions in the SDK module.

Reply:
so, wrapping it in `Seq(...)` here is just to coerce (effectively) `Seq[c.Expr[Attribute[_]]]` to `c.Expr[Seq[Attribute[_]]]`. at runtime, because varargs is already `Seq`, it just passes a type test and returns the same instance

Reply:
There are plenty of cases when we call `Attributes.fromSpecific` to build `Attributes`. For example:
1) https://github.com/typelevel/otel4s/blob/62835f7f6a1315572f7ae1b760e54c38c32b7d7e/core/metrics/src/main/scala/org/typelevel/otel4s/metrics/Measurement.scala#L56-L57
2) https://github.com/typelevel/otel4s/blob/62835f7f6a1315572f7ae1b760e54c38c32b7d7e/core/metrics/src/main/scala/org/typelevel/otel4s/metrics/ObservableMeasurement.scala#L33-L34
3) https://github.com/typelevel/otel4s/blob/62835f7f6a1315572f7ae1b760e54c38c32b7d7e/sdk/trace/src/main/scala/org/typelevel/otel4s/sdk/trace/SdkSpanBackend.scala#L101-L108

And so on.

[fromSpecific](https://github.com/typelevel/otel4s/blob/62835f7f6a1315572f7ae1b760e54c38c32b7d7e/core/common/src/main/scala/org/typelevel/otel4s/Attributes.scala#L189) is defined as the following:
```scala
attributes match {
  case a: Attributes => a
  case other         => (newBuilder ++= other).result()
}
```

If we wrap varargs into the `Seq` in the macro, there will be a second conversion: `(newBuilder ++= other).result()`. If we wrap varargs right into the `Attributes` the second conversion can be omitted. (unless I'm missing something)

Reply:
This is not related to the current scope, but should we check the size inside of the `fromSpecific`?

```diff
def fromSpecific(attributes: IterableOnce[Attribute[_]]): Attributes =
  attributes match {
    case a: Attributes                           => a
+   case it: immutable.Iterable[_] if it.isEmpty => Attributes.empty
    case other                                   => (newBuilder ++= other).result()
  }
```

Reply:
> If we wrap varargs into the `Seq` in the macro, there will be a second conversion: `(newBuilder ++= other).result()`. If we wrap varargs right into the `Attributes`

because varargs is already `Seq`, wrapping it in a call to `Seq.apply` (which just forwards to `Seq.from`) is a constant time operation and doesn't actually convert anything. from what I understand, the reason you can't directly convert `Seq[Expr[A]]` to `Expr[Seq[A]]` is because it... jumps compiler phases? I'm really not sure. but [it seems](https://stackoverflow.com/a/73860907/5101123) like wrapping it in a call to `Seq.apply` is the accepted way to sidestep the limitations of Scala 2 macros.

additionally, because Scala 3 macros are better, they just give you varargs as `Expr[Seq[A]]]`, so the Scala 2 macros have to also return `Seq` to have consistent behaviour.

> the second conversion can be omitted.

for the `oteljava` backend, the collection can be converted or appended to the java `Attributes` directly without being converted to scala `Attributes` first. additionally, if being appended to an existing set of `Attributes`, it would not need to be converted to `Attributes` first either, and doing so would be an unnecessary conversion

Reply:
> should we check the size inside of the `fromSpecific`?

probably? I can't imagine it making a huge difference, but also other collections seem to do it

> There are plenty of cases when we call `Attributes.fromSpecific` to build `Attributes`

this is a small nitpick, but I think it's generally more idiomatic to do `.to(Attributes)` (e.g. `List(Attribute("key", "value")).to(Attributes)`). it ends up still calling `fromSpecific`, but I think it's maybe a bit more readable? idk

apparently I missed `ObservableMeasurement` when searching through all the files. I opine that we should widen its overload from taking `Attributes` to taking `immutable.Iterable[Attribute[_]]`, both for consistency with the other APIs and for convenience (if passing a `Seq` that's not varargs, you don't need to do `: _*`)

> If we wrap varargs right into the `Attributes` the second conversion can be omitted.

regarding this: even if the operation would be O(1), you'd still need to do `.to(Attributes)` (or `Attributes.fromSpecific`), as the type is only statically known to be `immutable.Iterable[Attribute[_]]`, not `Attributes`

Reply:
Ah, I see. Thanks for the detailed explanation.